File: Languages/PHP/Day020.php
<!-- Languages - PHP - Day 20 -->
<!-- Foundational Concepts: Basics, syntax, loops, conditionals, functions, and variables. -->

<!-- Create a function that sorts an array using the Bubble Sort algorithm. -->


<?php

// Bubble Sort function
function bubbleSort($arr) {
    $n = count($arr);
    for ($i = 0; $i < $n - 1; $i++) {
        $swapped = false;
        for ($j = 0; $j < $n - $i - 1; $j++) {
            if ($arr[$j] > $arr[$j + 1]) {
                $temp = $arr[$j];
                $arr[$j] = $arr[$j + 1];
                $arr[$j + 1] = $temp;
                $swapped = true;
            }
        }
        if (!$swapped) break;
    }
    return $arr;
}

// Function to validate user input
function validateInput($numbers) {
    foreach ($numbers as $num) {
        if (!is_numeric(trim($num))) {
            return false;
        }
    }
    return true;
}

// Ask for user input
$numbers = readline("Enter numbers separated by commas: ");
$numbers = explode(",", $numbers);

// Validate user input
if (!validateInput($numbers)) {
    echo "Invalid input. Please enter only numbers separated by commas.\n";
    exit();
}

// Convert strings to integers
$numbers = array_map('intval', $numbers);

// Display original array
echo "Original array: ";
echo implode(" ", $numbers) . "\n";

// Sort the array
$sorted = bubbleSort($numbers);

// Print the sorted array
echo "Sorted array: ";
echo implode(" ", $sorted) . "\n";

?>

File: Languages/Go/Day020.go
// Languages - Go - Day 20
// Foundational Concepts: Basics, syntax, loops, conditionals, functions, and variables.

// Create a function that sorts an array using the Bubble Sort algorithm.


package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

// Bubble Sort function
func bubbleSort(arr []int) []int {
	for i := 0; i < len(arr); i++ {
		for j := 0; j < len(arr)-i-1; j++ {
			if arr[j] > arr[j+1] {
				arr[j], arr[j+1] = arr[j+1], arr[j]
			}
		}
	}
	return arr
}

// Function to read and validate integer input
func getValidatedInput(prompt string) int {
	reader := bufio.NewReader(os.Stdin)
	for {
		fmt.Print(prompt)
		input, _ := reader.ReadString('\n')
		input = strings.TrimSpace(input)
		number, err := strconv.Atoi(input)
		if err != nil {
			fmt.Println("Invalid input. Please enter a valid integer.")
			continue
		}
		return number
	}
}

func main() {
	// Get the number of elements from the user
	n := getValidatedInput("Enter the number of elements in the array: ")
	for n <= 0 {
		fmt.Println("Please enter a positive integer.")
		n = getValidatedInput("Enter the number of elements in the array: ")
	}

	// Get array elements from the user
	arr := make([]int, n)
	fmt.Println("Enter", n, "integers:")
	for i := 0; i < n; i++ {
		arr[i] = getValidatedInput(fmt.Sprintf("Element %d: ", i+1))
	}

	// Sort the array
	sortedArr := bubbleSort(arr)

	// Display the sorted array
	fmt.Println("\nSorted array is:", sortedArr)
}
